Strip color codes from ascii art line length calculations

Syphist opened the pull request:
## Description
This removes color and formatting codes from the line length calculation to allow for more control over custom ascii art. This would also fix issues I found when attempting to use the patch in #1220.

## Features
Allows for support for finer control of ascii art using escapes such as "[38;5;39m" for defining coloring and formatting in the terminal.

## Issues
None that I have found

## TODO
Nothing

## Fixed Issues
This so far only works if you directly paste the character for \033 or "" into the file. I was unable to get it to work for both.

Commit:
Using regex I was able to strip color codes from ascii art line length calculations.

Syphist commented:
I have made several more attempts to work around SC2001, but have had no success. I am wondering if it is too complex to do it with the search and replace in bash and may need to be done with sed. I will probably make further attempts at this, if anyone else has any ideas feel free to let me know too.

Commit:
Fixed SC2001 error, you also need to use \033 instead of pasting the character in.

Commit:
Added support for pasted character and removed unecessary variables.

Commit:
Made comment accurate

Syphist commented:
I have figured out the solution and have updated my code accordingly and have cleaned it up. I have tested both with the pasted character in the ascii file and the escape code and both work. This is also much faster than my initial sed implementation. I have updated the OP to reflect these changes as well as clarify a few things.

Commit:
Added patch from dylanaraps/neofetch/1220 as it runs faster in some instances and does not hinder performance where it doesn't work

Commit:
Fixed build error detected due to long comment.

Commit:
Merge branch 'master' of https://github.com/dylanaraps/neofetch into master
Merge all changes to master to this fork.

Commit:
Merge branch 'master' of https://github.com/dylanaraps/neofetch
Merge changes from main repo to this fork.

Commit:
Merge branch 'master' of https://github.com/dylanaraps/neofetch
Merge upstream changes.

Commit:
Merge remote-tracking branch 'upstream/master'
Merge upstream changes

Z-8Bit commented:
+1 I agree, it's much needed this.

Commit:
Merge branch 'dylanaraps:master' into master

Commit:
Merge branch 'dylanaraps:master' into master

thenick775 commented:
+1 this modification works well for me, have tested on Big Sur

![Screen Shot 2021-12-27 at 9 54 00 PM](https://user-images.githubusercontent.com/40526638/147532941-ec9313d1-0fdb-4469-b32b-910bbd5c6cc1.jpg)

Commit:
Merge branch 'dylanaraps:master' into master

fvegaswp commented:
Just tested this with colored ascii generated with lolcat and works great.

jonaramos commented:
Tested with colored ascii generated from jp2a and ascii-image-converter. Thanks Syphist.
![image](https://user-images.githubusercontent.com/19523697/178951364-ce0f42dd-70a9-4c46-849d-844d1cb05237.png)

hykilpikonna commented:
Lol I actually ran into this issue while creating HyFetch. My solution was to pass in a separate `ascii_len` variable:

![image](https://user-images.githubusercontent.com/22280294/184267880-708c0d1e-d429-4986-9b10-0a3c9c49323b.png)

Thanks a lot!

hykilpikonna commented:
Works!

![image](https://user-images.githubusercontent.com/22280294/184268065-556c907f-be28-4560-b6ce-67b68ca79213.png)

hykilpikonna commented:
Thank you for your contribution! 

This PR is [merged into hyfetch](https://github.com/hykilpikonna/hyfetch/commit/eeead8d7) since this repo (dylanaraps/neofetch) seems no longer maintained.

[HyFetch](https://github.com/hykilpikonna/hyfetch) is a fork of neofetch with LGBTQ pride flags, but the repo also maintains an updated version of the original neofetch, addressing many pull requests that are not merged in the original repo.

Read the ["Running Updated Original Neofetch" section](https://github.com/hykilpikonna/hyfetch#running-updated-original-neofetch) for more info!

hykilpikonna commented:
I found a bug in the last line of `strip_escape_codes`:

```diff
-     eval "$2=\"${output}\""
+     eval "$2='${output}'"
```

If there exists $ signs in the ascii art, this eval statement will replace the $ with a variable content. You should use single quotes to prevent this.

**Before Fix:**

![image](https://user-images.githubusercontent.com/22280294/186198566-121ad0f5-929a-418c-a756-38440e374dda.png)

**After Fix:**

![image](https://user-images.githubusercontent.com/22280294/186199833-1aa0b301-9f80-452b-8a0b-a629a9b3c36f.png)

hykilpikonna commented on neofetch:
```diff
@@ -4054,6 +4054,26 @@ image_backend() {
+        fi
+        output+="${char}"
+    done
+    eval "$2=\"${output}\""
```
```suggestion
    eval "$2='${output}'"
```

Commit:
Fix Bug With $ Character

Proposed by hykilpikonna. Fixes a bug that evals stuff that follows $ characters.

Co-authored-by: Hykilpikonna <me@hydev.org>

Syphist commented:
I have committed your fix. Thanks for finding a solution and testing further.

hykilpikonna commented:
Okay I found another bug lol

```diff
- while IFS=$'\n' read -r line; do
+ while IFS=$'\n' read line; do
```

The width is incorrectly calculated for ascii distros with backslashes in them (e.g. Bedrock) because the backslashes were not treated as escape characters in `read -r` while they are actually escaped when defining the ASCII:

![image](https://user-images.githubusercontent.com/22280294/188513474-d4d46523-b527-4ae5-b15f-407a282f97ca.png)

So, if `read -r` is used, the escaped backslash (`\\`) will not be resolved as one character but two characters, making the ascii length longer. (https://github.com/hykilpikonna/hyfetch/issues/19)

**Before Fix:**

![image](https://user-images.githubusercontent.com/22280294/188513693-5eecdd1a-e00a-4c00-96c4-e4dfb3d5eaaf.png)


**After Fix:**

![image](https://user-images.githubusercontent.com/22280294/188513667-96430ed1-1028-4808-a948-5d6748018c4a.png)

Commit:
Fix issue with backslashes

Syphist commented:
Thanks @hykilpikonna I fixed the issue on my fork as well just in case this ever gets merged upstream to neofetch or if someone bothers to use my fork.
